### PR TITLE
Improve error message when Navigator.pop() is called accidentally

### LINKED ISSUE #50258: Error message should be more verbose when Navigator.pop() is called accidentally
## Steps to Reproduce

1. Run `flutter create bug`.
2. Add `Navigator.pop()` to the `onPressed` method of the `FloatingActionButton`.
3. Run the app and press the button.
4. Hot reload the app.

**Expected results:** Verbose error message in the console that mentions that you may have accidentally called `Navigator.pop()` in the wrong context.

**Actual results:** 
`'package:flutter/src/widgets/navigator.dart': Failed assertion: line 2855 pos 12: '_history.isNotEmpty': is not true.`

<details>
  <summary>Logs</summary>

```
════════ Exception caught by widgets library ═══════════════════════════════════
The following assertion was thrown building Navigator-[GlobalObjectKey<NavigatorState> _WidgetsAppState#3f9ff](dirty, state: NavigatorState#69d35(tickers: tracking 0 tickers)):
'package:flutter/src/widgets/navigator.dart': Failed assertion: line 2855 pos 12: '_history.isNotEmpty': is not true.


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=BUG.md

The relevant error-causing widget was
    MaterialApp    lib/main.dart:XX
When the exception was thrown, this was the stack
#2      NavigatorState.build                package:flutter/…/widgets/navigator.dart:2855
#3      StatefulElement.build               package:flutter/…/widgets/framework.dart:4604
#4      ComponentElement.performRebuild     package:flutter/…/widgets/framework.dart:4492
#5      StatefulElement.performRebuild      package:flutter/…/widgets/framework.dart:4660
#6      Element.rebuild                     package:flutter/…/widgets/framework.dart:4216
...
════════════════════════════════════════════════════════════════════════════════
```
```
```

```
[✓] Flutter (Channel master, v1.14.7-pre.138, on Linux, locale en_US.UTF-8)
    • Flutter version 1.14.7-pre.138 at /.../flutter
    • Framework revision 49be146953 (5 hours ago), 2020-02-05 18:02:20 -0800
    • Engine revision 81dffd1241
    • Dart version 2.8.0 (build 2.8.0-dev.7.0 c8ed304e97)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /.../Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /snap/android-studio/81/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Linux toolchain - develop for Linux desktop
    • clang++ 9.0.0
    • GNU Make 4.2.1

[✓] Android Studio (version 3.5)
    • Android Studio at /snap/android-studio/81/android-studio
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] IntelliJ IDEA Ultimate Edition (version 2019.2)
    • IntelliJ at /.../idea-IU-192.6262.58
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (2 available)
    • ONEPLUS A6013 • 93a8e251 • android-arm64 • Android 10 (API 29)
    • Linux         • Linux    • linux-x64     • Linux

! Doctor found issues in 1 category.
```

</details>

### PR DESCRIPTION
**Pull request by JuYeong0413**
## Description

This pull request just added error message that `Navigator.pop` may have called accidentally.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/50258

## Tests

- `navigator_test.dart`
Added a test to check the error message is showing properly.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -4078,7 +4078,7 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
   @override
   Widget build(BuildContext context) {
     assert(!_debugLocked);
-    assert(_history.isNotEmpty);
+    assert(_history.isNotEmpty, 'You may have accidentally called Navigator.pop() in the wrong context.');
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/navigator.dart:4081]: This requires a test, you can in the test try to pop navigator to empty and catch the assertion error and verify its string matches expectation

Comment (JuYeong0413) [reply packages/flutter/lib/src/widgets/navigator.dart:4081]: I've added a test for this. Please take a look. Thank you! :)

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/navigator.dart:4081]: Looking at the error message again it feels a bit confusing. There are a lot of ways user can end up in this situation, and poping the wrong context is only one of them. If user arrive this assertion in a different way, this message can be misleading. It is too specific, I don't have a good idea on the top of my head too...

Comment (JuYeong0413) [reply packages/flutter/lib/src/widgets/navigator.dart:4081]: Hmm, I agree that this message is too specific.
```suggestion
    assert(_history.isNotEmpty, 'There is no navigation history in the current context. You may need to check Navigator routes.');
```
This is the only thing came into my mind... I wonder if this message would be helpful to users 🤔

**Comment by JuYeong0413** [comment]
@chunhtai I've updated the message, but I'm not very sure if this would be helpful to users. 🤔

**Comment by chunhtai** [comment]
@JuYeong0413 it does not looks it will add much information about the error, and i could not think of better message, because we cannot really tell what cause this assertion. I suggest we should instead try to catch this error in earlier stage and report the error there instead.

**Comment by JuYeong0413** [comment]
I was thinking about checking the `_history.length` when `Navigator.pop()` is called, and print the message there.
It's updated in https://github.com/flutter/flutter/pull/64409/commits/0fa3eb4c87b8ada4f122d58ba75681133569c8e2, but I couldn't find out how to check the `debugPrint` in the test. Is there any existing tests I could use an example to write the test about this?

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 2]
```diff
@@ -3835,6 +3835,9 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
       _flushHistoryUpdates(rearrangeOverlay: false);
       assert(entry.route._popCompleter.isCompleted);
     }
+    if (_history.length == 1) {
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/navigator.dart:3838]: This does not consider the case where there is localhistory in the route. where the pop will pop the localhistory but the route won't be popped

**Comment by goderbauer** [comment]
@JuYeong0413 Do you still have plans to follow-up on this one?

**Comment by JuYeong0413** [comment]
@goderbauer Sorry for the delay. I would like to keep working on this, but I don't have time to look at this for a few weeks and I felt I need to study much more about `Navigator` to improve the error message.

I'll close this for now and re-open when I'm ready. Thanks for the ping :-)

**Pull request closed without merging**